pkg/build/controller/controller.go

```diff
@@ -108,6 +163,44 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 	return nil
 }
 
+func (bc *BuildController) nextBuild(namespace, buildConfigName string) (*buildapi.Build, error) {
+	hasRunningBuilds := false
+	builds, err := buildutil.BuildConfigBuilds(bc.OSClient.Builds(namespace), buildConfigName, func(b buildapi.Build) bool {
+		// Detect if there are running builds
+		switch b.Status.Phase {
+		case buildapi.BuildPhaseRunning, buildapi.BuildPhasePending:
+			hasRunningBuilds = true
+			return false
+		}
+		// Only 'new' build can be scheduled to run next
+		return b.Status.Phase == buildapi.BuildPhaseNew
+	})
+	if hasRunningBuilds || err != nil {
+		return nil, err
+	}
+	var (
+		nextBuild           *buildapi.Build
+		previousBuildNumber int64
+	)
+	for i, b := range builds.Items {
+		buildNumber, err := buildutil.BuildNumber(&b)
+		if err != nil {
+			glog.Errorf("Error getting build number: %v", err)
+			continue
+		}
+		if previousBuildNumber == 0 {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+			continue
+		}
```

Comment:
(in fact even initializing it to zero is sufficient since we start our build numbers at 1 i believe)